File: UEPlugin/AgentBridge/Source/AgentBridge/Private/UATRunner.cpp
// UATRunner.cpp
// AGENT + UE5 可操作层 - UAT 子进程封装实现

#include "UATRunner.h"

#include "HAL/PlatformProcess.h"
#include "Misc/Paths.h"

FUATRunner::FUATRunner()
{
	RunUATPath = DetectRunUATPath();
}

FString FUATRunner::DetectRunUATPath() const
{
	const FString EngineDir = FPaths::EngineDir();

#if PLATFORM_WINDOWS
	FString Candidate = EngineDir / TEXT("Build/BatchFiles/RunUAT.bat");
#else
	FString Candidate = EngineDir / TEXT("Build/BatchFiles/RunUAT.sh");
#endif

	Candidate = FPaths::ConvertRelativePathToFull(Candidate);
	if (!FPaths::FileExists(Candidate))
	{
		return FString();
	}
	return Candidate;
}

bool FUATRunner::IsUATAvailable() const
{
	return !RunUATPath.IsEmpty() && FPaths::FileExists(RunUATPath);
}

FUATRunResult FUATRunner::ExecuteUAT(const FString& Args, const bool bSync)
{
	FUATRunResult Result;
	Result.CommandLine = FString::Printf(TEXT("\"%s\" %s"), *RunUATPath, *Args);

	if (!IsUATAvailable())
	{
		Result.ErrorMessage = FString::Printf(TEXT("RunUAT not found: %s"), *RunUATPath);
		UE_LOG(LogTemp, Error, TEXT("[AgentBridge UAT] %s"), *Result.ErrorMessage);
		return Result;
	}

	uint32 ProcessId = 0;
	void* StdOutReadPipe = nullptr;
	void* StdOutWritePipe = nullptr;
	void* StdErrReadPipe = nullptr;
	void* StdErrWritePipe = nullptr;

	if (bSync)
	{
		FPlatformProcess::CreatePipe(StdOutReadPipe, StdOutWritePipe);
		FPlatformProcess::CreatePipe(StdErrReadPipe, StdErrWritePipe);
	}

	UE_LOG(LogTemp, Log, TEXT("[AgentBridge UAT] Launch: %s"), *Result.CommandLine);

	FProcHandle ProcHandle;
#if PLATFORM_WINDOWS
	ProcHandle = FPlatformProcess::CreateProc(
		*RunUATPath,
		*Args,
		/*bLaunchDetached=*/true,
		/*bLaunchHidden=*/true,
		/*bLaunchReallyHidden=*/false,
		&ProcessId,
		/*PriorityModifier=*/0,
		nullptr,
		bSync ? StdOutWritePipe : nullptr,
		nullptr,
		bSync ? StdErrWritePipe : nullptr);
#else
	ProcHandle = FPlatformProcess::CreateProc(
		*RunUATPath,
		*Args,
		/*bLaunchDetached=*/true,
		/*bLaunchHidden=*/true,
		/*bLaunchReallyHidden=*/false,
		&ProcessId,
		/*PriorityModifier=*/0,
		nullptr,
		bSync ? StdOutWritePipe : nullptr,
		nullptr);
#endif

	if (!ProcHandle.IsValid())
	{
		Result.ErrorMessage = FString::Printf(TEXT("Failed to launch UAT process: %s"), *RunUATPath);
		UE_LOG(LogTemp, Error, TEXT("[AgentBridge UAT] %s"), *Result.ErrorMessage);
		if (StdOutReadPipe || StdOutWritePipe) FPlatformProcess::ClosePipe(StdOutReadPipe, StdOutWritePipe);
		if (StdErrReadPipe || StdErrWritePipe) FPlatformProcess::ClosePipe(StdErrReadPipe, StdErrWritePipe);
		return Result;
	}

	Result.bLaunched = true;

	if (!bSync)
	{
		Result.bCompleted = false;
		Result.ExitCode = -1;
		UE_LOG(LogTemp, Log, TEXT("[AgentBridge UAT] Process launched in async mode, PID=%u"), ProcessId);
		return Result;
	}

	// 同步模式：持续读管道，避免大输出导致管道阻塞。
	while (FPlatformProcess::IsProcRunning(ProcHandle))
	{
		if (StdOutReadPipe)
		{
			Result.StdOut += FPlatformProcess::ReadPipe(StdOutReadPipe);
		}
		if (StdErrReadPipe)
		{
			Result.StdErr += FPlatformProcess::ReadPipe(StdErrReadPipe);
		}
		FPlatformProcess::Sleep(0.05f);
	}

	FPlatformProcess::WaitForProc(ProcHandle);

	if (StdOutReadPipe)
	{
		Result.StdOut += FPlatformProcess::ReadPipe(StdOutReadPipe);
	}
	if (StdErrReadPipe)
	{
		Result.StdErr += FPlatformProcess::ReadPipe(StdErrReadPipe);
	}

	int32 ReturnCode = -1;
	if (!FPlatformProcess::GetProcReturnCode(ProcHandle, &ReturnCode))
	{
		ReturnCode = -1;
	}

	Result.bCompleted = true;
	Result.ExitCode = ReturnCode;

	if (StdOutReadPipe || StdOutWritePipe) FPlatformProcess::ClosePipe(StdOutReadPipe, StdOutWritePipe);
	if (StdErrReadPipe || StdErrWritePipe) FPlatformProcess::ClosePipe(StdErrReadPipe, StdErrWritePipe);
	FPlatformProcess::CloseProc(ProcHandle);

	UE_LOG(LogTemp, Log, TEXT("[AgentBridge UAT] Sync finished, ExitCode=%d"), ReturnCode);
	return Result;
}

FUATRunResult FUATRunner::BuildCookRun(
	const FString& Platform,
	const FString& Configuration,
	const bool bSync)
{
	const FString ProjectPath = FPaths::GetProjectFilePath();
	const FString Args = FString::Printf(
		TEXT("BuildCookRun -project=\"%s\" -platform=%s -clientconfig=%s -cook -stage -pak"),
		*ProjectPath,
		*Platform,
		*Configuration);
	return ExecuteUAT(Args, bSync);
}

FUATRunResult FUATRunner::RunAutomationTests(
	const FString& Filter,
	const FString& ReportPath,
	const bool bSync)
{
	const FString ProjectPath = FPaths::GetProjectFilePath();
	// UE5.5 不存在独立的 RunAutomationTests 子命令，需通过 BuildCookRun 的 editortest 路径触发。
	FString Args = FString::Printf(
		TEXT("BuildCookRun -project=\"%s\" -run -editortest -unattended -nullrhi -NoP4"),
		*ProjectPath);

	if (!Filter.IsEmpty())
	{
		Args += FString::Printf(TEXT(" -RunAutomationTest=%s"), *Filter);
	}
	else
	{
		Args += TEXT(" -RunAutomationTests");
	}

	// BuildCookRun 没有稳定的 -ReportOutputPath 参数入口，暂保留字段用于上层记录。
	if (!ReportPath.IsEmpty())
	{
		Args += FString::Printf(TEXT(" -addcmdline=\"-ReportExportPath=\\\"%s\\\"\""), *ReportPath);
	}

	return ExecuteUAT(Args, bSync);
}

FUATRunResult FUATRunner::RunGauntlet(
	const FString& TestConfigName,
	const bool bSync)
{
	const FString ProjectPath = FPaths::GetProjectFilePath();
	const FString Args = FString::Printf(
		TEXT("RunGauntlet -project=\"%s\" -Test=%s -unattended"),
		*ProjectPath,
		*TestConfigName);
	return ExecuteUAT(Args, bSync);
}

FUATRunResult FUATRunner::RunCustomCommand(
	const FString& Command,
	const bool bSync)
{
	FString Args = Command.TrimStartAndEnd();
	if (Args.IsEmpty())
	{
		FUATRunResult Result;
		Result.ErrorMessage = TEXT("RunCustomCommand requires non-empty command");
		return Result;
	}

	if (!Args.Contains(TEXT("-project="), ESearchCase::IgnoreCase))
	{
		Args += FString::Printf(TEXT(" -project=\"%s\""), *FPaths::GetProjectFilePath());
	}

	return ExecuteUAT(Args, bSync);
}

File: Scripts/bridge/uat_runner.py
"""
uat_runner.py
=============
AGENT + UE5 可操作層 — UAT (Unreal Automation Tool) Python 封装。

UE5 官方模块：UAT

本模块提供从 Python 端调用 UAT 的能力，供 Orchestrator 和 CLI 使用。
与 C++ 端的 FUATRunner 功能对等，但运行在 Editor 进程外部。

使用场景：
  - Python CLI（agent-ue5 build / agent-ue5 test）
  - Python Orchestrator 中触发构建
  - CI/CD 脚本

注意：UAT 运行在引擎进程外部（C# 程序），
本模块通过 subprocess 启动 UAT 进程。
"""

from __future__ import annotations

import os
import platform
import subprocess
from dataclasses import dataclass

try:
    from project_config import get_saved_dir, get_uproject_path
except ImportError:  # pragma: no cover - 兼容包导入
    from .project_config import get_saved_dir, get_uproject_path


# ============================================================
# 运行结果
# ============================================================

@dataclass
class UATRunResult:
    """UAT 子进程运行结果。对应 C++ 端的 FUATRunResult。"""
    launched: bool = False
    completed: bool = False
    exit_code: int = -1
    command_line: str = ""
    stdout: str = ""
    stderr: str = ""
    error_message: str = ""

    @property
    def is_success(self) -> bool:
        return self.launched and self.completed and self.exit_code == 0


# ============================================================
# UAT Runner
# ============================================================

class UATRunner:
    """
    UAT 子进程封装（Python 版）。
    对应 C++ 端的 FUATRunner。
    """

    def __init__(self, project_path: str = "", engine_dir: str = ""):
        """
        初始化 UAT Runner。

        参数：
            project_path: .uproject 文件路径（空=自动检测）
            engine_dir: 引擎目录（可传 UE 安装根目录，或直接传 .../Engine 目录）
        """
        self.project_path = project_path or str(get_uproject_path())
        self.engine_dir = engine_dir
        self._run_uat_path = self._detect_run_uat_path()

    # ============================================================
    # 核心操作
    # ============================================================

    def build_cook_run(
        self,
        platform_name: str = "Win64",
        configuration: str = "Development",
        sync: bool = True,
    ) -> UATRunResult:
        """
        执行 BuildCookRun。
        对应 UAT 命令：RunUAT BuildCookRun -project=... -build -cook -stage -pak
        """
        args = (
            f"BuildCookRun"
            f' -project="{self.project_path}"'
            f" -platform={platform_name}"
            f" -clientconfig={configuration}"
            f" -build -cook -stage -pak"
            f" -unattended -utf8output"
        )
        return self._execute_uat(args, sync=sync)

    def run_automation_tests(
        self,
        test_filter: str = "Project.AgentBridge",
        report_path: str = "",
        sync: bool = True,
    ) -> UATRunResult:
        """
        通过 UAT 运行自动化测试。
        UE5.5 中不再依赖独立的 RunAutomationTests 子命令，
        统一走 BuildCookRun 的 editortest 路径：
        RunUAT BuildCookRun -project=... -run -editortest -RunAutomationTest=...
        """
        args = (
            f"BuildCookRun"
            f' -project="{self.project_path}"'
            f" -run -editortest -unattended -utf8output -nullrhi -NoP4"
        )
        if test_filter:
            args += f" -RunAutomationTest={test_filter}"
        else:
            args += " -RunAutomationTests"

        if report_path:
            resolved_report_path = self._resolve_report_path(report_path)
            # BuildCookRun 没有稳定的直接报告参数，这里通过附加命令行透传给 Automation。
            args += f' -addcmdline="-ReportExportPath=\\"{resolved_report_path}\\""'
        return self._execute_uat(args, sync=sync)

    def run_gauntlet(
        self,
        test_config_name: str = "AgentBridge.AllTests",
        sync: bool = True,
    ) -> UATRunResult:
        """
        启动 Gauntlet 测试会话。
        对应 UAT 命令：RunUAT RunGauntlet -Test=... -project=...
        """
        args = (
            f"RunGauntlet"
            f' -project="{self.project_path}"'
            f" -Test={test_config_name}"
            f" -unattended -utf8output"
        )
        return self._execute_uat(args, sync=sync)

    def run_custom(self, uat_command: str, sync: bool = True) -> UATRunResult:
        """执行任意 UAT 命令。"""
        args = f'{uat_command} -project="{self.project_path}" -unattended -utf8output'
        return self._execute_uat(args, sync=sync)

    # ============================================================
    # 状态检查
    # ============================================================

    @property
    def is_available(self) -> bool:
        """检查 RunUAT 是否存在。"""
        return self._run_uat_path != "" and os.path.isfile(self._run_uat_path)

    @property
    def run_uat_path(self) -> str:
        """获取 RunUAT 路径。"""
        return self._run_uat_path

    # ============================================================
    # 内部方法
    # ============================================================

    def _execute_uat(self, args: str, sync: bool = True) -> UATRunResult:
        """执行 UAT 子进程。"""
        result = UATRunResult()

        if not self.is_available:
            result.error_message = f"RunUAT not found at: {self._run_uat_path}"
            return result

        full_command = f'"{self._run_uat_path}" {args}'
        result.command_line = full_command

        try:
            if sync:
                proc = subprocess.run(
                    full_command,
                    shell=True,
                    capture_output=True,
                    text=True,
                    timeout=3600,  # 1 小时超时
                )
                result.launched = True
                result.completed = True
                result.exit_code = proc.returncode
                result.stdout = proc.stdout
                result.stderr = proc.stderr
            else:
                proc = subprocess.Popen(
                    full_command,
                    shell=True,
                    stdout=subprocess.DEVNULL,
                    stderr=subprocess.DEVNULL,
                )
                result.launched = True
                result.completed = False
                result.exit_code = -1

        except subprocess.TimeoutExpired:
            result.launched = True
            result.completed = False
            result.error_message = "UAT process timed out after 3600 seconds"
        except Exception as e:
            result.error_message = f"Failed to launch UAT: {e}"

        return result

    def _detect_run_uat_path(self) -> str:
        """自动检测 RunUAT 路径。"""
        if self.engine_dir:
            base = self.engine_dir
        else:
            # 尝试从环境变量获取
            base = os.environ.get("UE_ENGINE_DIR", "")

        if not base:
            return ""

        if platform.system() == "Windows":
            candidate_paths = [
                os.path.join(base, "Build", "BatchFiles", "RunUAT.bat"),
                os.path.join(base, "Engine", "Build", "BatchFiles", "RunUAT.bat"),
            ]
        else:
            candidate_paths = [
                os.path.join(base, "Build", "BatchFiles", "RunUAT.sh"),
                os.path.join(base, "Engine", "Build", "BatchFiles", "RunUAT.sh"),
            ]

        for path in candidate_paths:
            if os.path.isfile(path):
                return os.path.abspath(path)

        return ""

    def _resolve_report_path(self, report_path: str) -> str:
        """将相对测试报告路径解析到项目 Saved/ 下。"""
        if os.path.isabs(report_path):
            return report_path
        return str(get_saved_dir() / report_path)

